Add BlendSpace BlendPoint weight transitions by velocity limiting

ZeEndy opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->
Closes godotengine/godot-proposals#12335 
Still need more input on the proposal to improve usability but its pretty much done for the most part. This is my first PR did the best I could but I'm sure some things can defiantly be improved.

Implementation of blend point weight smoothing

- [x] Weight smoothing to AnimationBlendSpace2D
- [x] Weight smoothing to AnimationBlendSpace1D
- [x] Editor UI (change smoothing to only appear when using sync and continuous)

[Demo project file](https://github.com/ZeEndy/BlendSpace-BlendPoint-Smoothing-Demo.git)

ZeEndy committed:
Add BlendSpace BlendPoint weight velocity limiting

ZeEndy commented:
I'm being dumb ignore

TokageItLab commented:
It looks like you have multiple implementations mixed up (weight implementation and smooth implementation can be separated). But from a quick look, I think it makes sense to have the BlendPoint have weight.

However, I am quite skeptical of the need for time-dependent smoothing using `move_toward()`. This should obviously be done before passing Vector2 to BlendSpace2D and be sufficient.

If you want to adjust the slope of the blendmap, you should not use `move_toward()`, but rather a mathematical implementation that maps the linear interplation to a smoothstep or tanh S-curve and adjusts the rapidity of that slope (like adjusting the contrast of an image).

ZeEndy commented:
> However, I am quite skeptical of the need for time-dependent smoothing using `move_toward()`. This should obviously be done before passing Vector2 to BlendSpace2D and be sufficient.

I changed the delta time to not be based on the playback's delta time and updated the demo to be playable, If you switch with tab while playing you can see there is a major difference between smoothness when just interpolating the blend_point vector2 and using this method. It is usually common practice to have a blendspace to have 1 axis responsible for rotation and the other for speed, If you tried to interpolate that the character would have to go through each rotation before being able animate in a different direction even if you delay the interpolation of speed.

This is why this PR and feature proposal was made, due to a lack of a way to actually smoothly move between 2 radically different (in position) points with many points in between inside the blendspace with delta, without having to remake the blendspace in a script with way too many parameters to take care of or remake the entire blendspace system in gdextension to achieve the same effect as shown in the demo.

I dont mind making this a different animation node but I think there is a massive misunderstanding between the 2 methods of smoothing. I've stated in the proposal before why the method you suggested is not sufficient when it comes to large blendspaces. I've tried said method on multiple projects in different configurations and in none of them was it good enough as it still required the ``blend_position`` to cross unnecessary animations points causing jank.

I can however agree that just using ``move_toward()`` is not the solution and it would be better to figure out what would be a good way to allow users to modify the smoothing calculation themselves or what options could be included that would suffice for most users.


> Or, if it matches the your usage, I think we could welcome the approach of adding a new AnimationNode like GradientSpace2D that places a number of circles with blur.

I'm not sure about the specific implementation of this animation node as I dont know if the gradientblenspace2d with a bunch of circles with a blur would be a good idea. Maybe if it was made to look like heatmap that changes color depending on the weight of the animation, but I'd have to look into to how to achieve that effect while clamped to the the blendspace "mesh" that wouldn't be insanely expensive.
Heres a quick mock up of where a user would have made the blend_position quickly move left to right. 
![image](https://github.com/user-attachments/assets/d4a9e849-21bf-406f-8917-7898fa145fa9)

TokageItLab commented:
For explanation, first assume that the weights are all 1's to simplify the calculation.

Consider the current point P in triangle ABC.

Draw a line from vertex A through the current point to intersect line BC. The intersection with line BC is A', and the weight of the A can be calculated as follows:

```weight_A = remap(A.distance_to(P), 0, A.distance_to(A'), 0, 1)```

Here the weights are linearly interpolated in the range 0 to 1. Then, you can remap it to an arbitrary curve completion and you should be able to remove the move_toward and time base delta.

ZeEndy commented:
I'm not removing delta since thats not the point of the PR. The point of the pr isn't to change how the weights are applied across 3 points in a triangle, the point of the PR is across the entire blendspace. I'm not here to change the curvature of weights I'm here to change the fact there is no way to quickly transition from triangle ABC to triangle GHI without going through DEF and applying the weights fully causing animation jank. 

Smoothing the ``blend_position`` Vector2 wont fix that and its not sufficient at all. If you try demo(or just look at the video) you'll see I put in both your suggested method from before with the Vector2 and my smoothing that you can hot swap with mine on or off while walking around and its a night and day difference.

https://github.com/user-attachments/assets/96db3d4c-8b52-41f3-9af8-b0c837c05ca2

TokageItLab commented:
So are you saying that you want to discretely divide the triangle and crossfade the invalid / valid points as the point cross the boundary? If so, delta might make sense, but at least it should be a property called transition duration or blend time, not smooth.

ZeEndy commented:
Yea it's something among those lines. DC can definitely be done as well I just did continuas as that seemed like the logical first place. I only named it smoothing since that's what it was called in UE and I have no real other reference for such a feature. I'll definitely look into implementing your suggestions when I can.

TokageItLab commented on scene/animation/animation_blend_space_2d.cpp:
```diff
@@ -510,28 +554,56 @@ AnimationNode::NodeTimeInfo AnimationNodeBlendSpace2D::_process(const AnimationM
+
+		int mind_idx = 0;
+		Engine *eng = Engine::get_singleton();
+		double delta = static_cast<float>(override_delta ? (eng->get_process_step() * eng->get_time_scale()) : pi.delta);
```
If you want to do this, we may need to add a separate PR to add it to AnimationTree or AnimationMixer.

ZeEndy replied:
By default the timescale node does affect the blendspace but in my own use case I've found that want to decouple the velocity limiting from speed from the TimeScale. 
![image](https://github.com/user-attachments/assets/7083654f-75a6-4ca8-811c-5c265945d214)
I'll look into another way to do it so that its based the animationtree's advance function as otherwise if we're to use just the default playback delta that will make it impossible for the blendspace to change weights while using velocity limit and timescale is set to 0.0 even if the velocity limit is multiplied to be higher.

TokageItLab replied:
Do you want to bypass the TimeScale and not scale the BlendSpace delta itself, but only change the TimeScale of the child animations?

I understand the demand for this, but I think it should be a separate PR. I remember we had a discussion somewhere before about having TimeScale as a parameter for all AnimationNodes that have ports.

So it should apply to BlendSpace as well as other AnimationNodes at the same time.

ZeEndy replied:
> I understand the demand for this, but I think it should be a separate PR. I remember we had a discussion somewhere before about having TimeScale as a parameter for all AnimationNodes that have ports.

I did see that in the comments while looking around. It could be just a temporary bool that's optional for now while that hasn't been implemented yet, since 5.0 is quite far away right now. I'd go back and change it in a future pr to fall more in line with the rest when it's in.
Otherwise the blendspace wont be able to change the weights at all while timescale is set 0.0 while without velocity limit it can which doesn't make sense.

TokageItLab replied:
The workaround at the moment is to make every point placed in BlendSpace a nested BlendTree instead of an AnimationNode, and place a TimeScale within the nest for each point, and change all TimeScales simultaneously by script. But I know this is tedious, so which is why we had the discussion like above.

ZeEndy replied:
Thats true but for now for the sake of convenience and sanity of anyone trying to use this it would be beneficial to keep it like this (as going through all the blend points' blendtree's in a script would be a nightmare). At a later point when said time scale gets added to the base animation node then it in my opinion it could be made so that the blandspace has 2 timescales 1 that's default and takes care of the child animation nodes and velocity limit and an override time scale that takes care of the velocity limit only if enabled.

TokageItLab replied:
No. I disagree with override_delta on consistency and maintainability. When BlendSpace delta is scaled, it should simply multiplys velocity, which is consistent with other transition nodes and etc.

If you add a property for a temporary workaround as a band-aid, there will be a breakdown in compatibility when we later deprecate that property by appropreate way such as implement base AnimationNode's TimeScale. So it should not be done, and we will not be able to approve it,

ZeEndy replied:
``override_delta`` has been removed. I sure as hell don't agree on having a blend tree per blend point as good workaround but there is no good way to solve this at the moment. At least in a way we both can agree on.

TokageItLab commented on scene/animation/animation_blend_space_1d.cpp:
```diff
@@ -347,18 +369,48 @@ AnimationNode::NodeTimeInfo AnimationNodeBlendSpace1D::_process(const AnimationM
+
+		int mind_idx = 0;
+		Engine *eng = Engine::get_singleton();
+		double delta = eng->get_process_step() * eng->get_time_scale();
```
```suggestion
		double delta = p_playback_info.delta;
```
Delta should be retrieved from time info to make consistent for all other AnimationNodes.

ZeEndy replied:
Yep I was working on 2D the entire time and I wasn't planning on doing 1D till I had it at a point where I can pretty much transfer the changes. I forgot to make it a draft again my bad.

TokageItLab commented:
Once I made it a draft, so please mark as ready to review when finish working. If you have any other questions, you can menshon or dm me.
